Data/ByteString/Base64/Internal.hs

```diff
@@ -121,51 +121,46 @@ mkEncodeTable alphabet@(PS afp _ _) =
     table = B.pack $ concat $ [ [ix j, ix k] | j <- [0..63], k <- [0..63] ]
 
 -- | Decode a base64-encoded string.  This function strictly follows
--- the specification in
--- <http://tools.ietf.org/rfc/rfc4648 RFC 4648>.
+-- the specification in <http://tools.ietf.org/rfc/rfc4648 RFC 4648>.
+--
 -- This function takes the decoding table (for @base64@ or @base64url@) as
--- the first paramert.
+-- the first parameter.
+--
+-- For validation of padding properties, see note: $Validation
+--
 decodeWithTable :: Padding -> ForeignPtr Word8 -> ByteString -> Either String ByteString
 decodeWithTable _ _ (PS _ _ 0) = Right B.empty
-decodeWithTable padding decodeFP bs@(PS !fp !o !l) = unsafePerformIO $
+decodeWithTable padding decodeFP bs =
    case padding of
      Padded
-       | r == 1 -> err "Base64-encoded bytestring has invalid size"
-       | r /= 0 -> err "Base64-encoded bytestring required to be padded"
-       | otherwise -> go bs
+       | r == 1 -> Left "Base64-encoded bytestring has invalid size"
+       | r /= 0 -> Left "Base64-encoded bytestring required to be padded"
+       | otherwise -> unsafePerformIO $ go bs
      Don'tCare
-       | r == 0 -> go bs
-       | r == 2 -> go (B.append bs (B.replicate 2 0x3d))
-       | r == 3 -> go (B.append bs (B.replicate 1 0x3d))
-       | otherwise -> err "Base64-encoded bytestring has invalid size"
+       | r == 0 -> unsafePerformIO $ go bs
+       | r == 2 -> unsafePerformIO $ go (B.append bs (B.replicate 2 0x3d))
+       | r == 3 -> validateLastPad bs invalidPad $ go (B.append bs (B.replicate 1 0x3d))
+       | otherwise -> Left "Base64-encoded bytestring has invalid size"
      Unpadded
-       | r == 0 -> validateUnpadded (go bs)
-       | r == 2 -> validateUnpadded (go (B.append bs (B.replicate 2 0x3d)))
-       | r == 3 -> validateUnpadded (go (B.append bs (B.replicate 1 0x3d)))
-       | otherwise -> err "Base64-encoded bytestring has invalid size"
+       | r == 0 -> validateLastPad bs noPad $ go bs
+       | r == 2 -> validateLastPad bs noPad $ go (B.append bs (B.replicate 2 0x3d))
+       | r == 3 -> validateLastPad bs noPad $ go (B.append bs (B.replicate 1 0x3d))
```

Comment:
the logic for validation has been moved to the standalone function `validateLastPad`